bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/RepositoryImpl.java

```diff
@@ -0,0 +1,147 @@
+ *
+ * @since 2201.9.0
+ */
+public class RepositoryImpl implements Repository {
+
+    private static final Map<ObjectValue, ObjectValue> serviceListenerMap = new HashMap<>();
+    private static final Map<ObjectValue, ObjectValue> listenerServiceMap = new HashMap<>();
+    private static final String nodeId = generateNodeId();
+    private static String balHome;
+    private static String balVersion;
+
+    @Override
+    public List<Artifact> getArtifacts() {
+        List<Artifact> artifacts = new ArrayList<>();
+        for (Map.Entry<ObjectValue, ObjectValue> entry : serviceListenerMap.entrySet()) {
+            ObjectValue service = entry.getKey();
+            ObjectValue listener = entry.getValue();
+            Artifact artifact = createArtifact(service, listener);
+            artifacts.add(artifact);
+        }
+
+        for (Map.Entry<ObjectValue, ObjectValue> entry : listenerServiceMap.entrySet()) {
+            ObjectValue listener = entry.getKey();
+            ObjectValue service = entry.getValue();
+
+            if (!serviceListenerMap.containsKey(service)) {
+                Artifact artifact = createArtifact(service, listener);
+                artifacts.add(artifact);
+            }
+        }
+
+        return artifacts;
+    }
+
+    @Override
+    public Node getNode() {
+        return new NodeImpl(nodeId, balVersion, balHome, System.getProperty("os.name"),
+                System.getProperty("os.version"));
+    }
+
+    private Artifact createArtifact(ObjectValue service, ObjectValue listener) {
+        ArtifactImpl artifact = new ArtifactImpl(service.toString(), Artifact.ArtifactType.SERVICE);
+        List<ObjectValue> listeners = (List<ObjectValue>) artifact.getDetail("listeners");
+        if (listeners == null) {
+            listeners = new ArrayList<>();
+            artifact.addDetail("listeners", listeners);
+        }
+        listeners.add(listener);
+        BServiceType serviceType = (BServiceType) TypeUtils.getImpliedType(service.getOriginalType());
+        artifact.addDetail("attachPoint", serviceType.attachPoint);
+        artifact.addDetail("resources", serviceType.getResourceMethods());
+        artifact.addDetail("remotes", serviceType.getRemoteMethods());
+        return artifact;
+    }
+
+    public static void addServiceListener(BObject listener, BObject service, Object attachPoint) {
+        BServiceType serviceType = (BServiceType) service.getType();
+        serviceType.attachPoint = attachPoint;
+        serviceListenerMap.put((ObjectValue) service, (ObjectValue) listener);
+        listenerServiceMap.put((ObjectValue) listener, (ObjectValue) service);
+    }
+
+    public static void addBallerinaInformation(String balHome, String balVersion) {
+        RepositoryImpl.balHome = balHome;
+        RepositoryImpl.balVersion = balVersion;
+    }
+
+    private static String generateNodeId() {
+        return UUID.randomUUID().toString();
```

Comment:
Are we just going to return the uuid string for this?

Reply:
We can add a prefix too. Like `balNode-`. @warunalakshitha WDYT?

Reply:
As per the offline discussion with @warunalakshitha, we decided to only provide UUID for node id and let the vendor decide upon the node prefix. For example, MI integration control plane can configure it from their side with something like `dev_node`.